src/grafana_agent.py

```diff
@@ -198,23 +258,81 @@ def __init__(self, *args):
                     self.framework.observe(self.on[outgoing].relation_joined, self._update_status)
                     self.framework.observe(self.on[outgoing].relation_broken, self._update_status)
 
+    def _get_tracing_receiver_url(self, protocol: ReceiverProtocol):
+        scheme = "http"
+        try:
+            if self._charm.cert.available:  # type: ignore
+                scheme = "https"
+        except AttributeError:
+            pass
+
+        # assume we're doing this in-model, since this charm doesn't have ingress
+        if receiver_protocol_to_transport_protocol[protocol] == TransportProtocolType.grpc:
+            return f"{socket.getfqdn()}:{self._tracing_receivers_ports[protocol]}"
+        return f"{scheme}://{socket.getfqdn()}:{self._tracing_receivers_ports[protocol]}"
+
+    @property
+    def _force_enabled_tracing_protocols(self) -> Set[ReceiverProtocol]:
+        """Return a list of tracing receivers that have been force-enabled (by config)."""
+        return {
+            receiver
+            for receiver in get_args(ReceiverProtocol)
+            if self.config.get(f"always_enable_{receiver}")
+        }
+
+    @property
+    def _requested_tracing_protocols(self) -> Set[ReceiverProtocol]:
+        """All receiver protocols that have been requested by our related apps."""
+        return set(self._tracing_provider.requested_protocols()).union(
+            self._force_enabled_tracing_protocols
+        )
+
+    def _update_tracing_provider(self):
+        self._tracing_provider.publish_receivers(
+            tuple(
+                (protocol, self._get_tracing_receiver_url(protocol))
+                for protocol in self._requested_tracing_protocols
+            )
+        )
+
     def _on_cert_changed(self, _event):
         """Event handler for cert change."""
         self._update_config()
         self._update_ca()
         self._update_status()
+        self._update_tracing_provider()
+
+    def _on_tracing_endpoint_changed(self, _event) -> None:
+        """Event handler for the tracing endpoint-changed event."""
+        self._update_config()
+        self._update_status()
+        self._update_tracing_provider()
 
-    def _on_mandatory_relation_event(self, _event=None):
-        """Event handler for any mandatory relation event."""
+    def _on_tracing_endpoint_removed(self, _event) -> None:
+        """Event handler for the tracing endpoint-removed event."""
         self._update_config()
         self._update_status()
+        self._update_tracing_provider()
```

Comment:
Idea: we seem to be having this sequence of `_update_config()`, `_update_status()`, `_update_tracing_provider()` in several following methods. Could we extract it to a single update method that would be calling all that?

Reply:
+1 to that, but refactoring this code is out of scope for this PR. 
I'd like to move the status management in `collect-unit-status` where it should be, and then indeed, update_config should take care of updating all configs. I didn't want to break with the existing pattern for this specific PR